File: GFG/Count the Substring/Count the Substring.java
class Solution {
    long countSubstring(String S){
        
        int n=S.length();
        long[]dp=new long[2*n+1];// no. of substring with >= zeroes
        int cnt=n;
        int minus=0;
        for(int i=0;i<n;i++){
            if(S.charAt(i)=='0'){
                cnt--;
            }else{
                cnt++;
            }
            if(cnt<=n){
                minus++;
            }
            dp[cnt]++;
        }
        
        
        long ans=0L;
        int idx=n;
        for(int i=0;i<n;i++){
           ans+=(n-i - minus)*1l;
           if(S.charAt(i)=='1'){
               idx++;
               dp[idx]--;
               minus+=dp[idx];
           }else{
               minus--;
               minus-=dp[idx];
               idx--;
               dp[idx]--;
           }
        }
        return ans;
    }
}

File: GFG/Count the Substring/Count the Substring.cpp
class Solution{
public:
  long long countSubstring(string s){
    // start writing from here
    long long ans=0;
    unordered_map<long long,long long> help; // start unodering     
    vector<long long> dp(s.size(),0); // define the vector size
    long long sum=-1;
    if(s[0]=='1'){
         dp[0]=1;
         ans=1;
         sum=1;
    }
    help[sum]=0;
    help[0]=-1;
    for(int i=1;i<s.size();i++){
        long long temp=0;
        if(s[i]=='0')
            sum-=1;
        else
            sum+=1;
        if(s[i]=='0'){
            if(help.find(sum)!=help.end()) {
                if(help[sum]!=-1){
                    temp+=dp[help[sum]];
                }
            }
        }
        else{
            if(help.find(sum)!=help.end()){
                if(help[sum]!=-1){
                    temp=dp[help[sum]]+i-help[sum]-1;
                }
                else{
                    temp=i-help[sum]-1;
                }
            }
            else{
                temp+=i+1;
            }
        }
        help[sum]=i;
        dp[i]=temp;
        ans+=temp;
    }
    return ans;
  }
};
